### PR TITLE
Add Positional Encoding

### PR DESCRIPTION
**Pull request by yiweny**
@weihua916 mentioned we want to modularize the positional encoding to be used by both numerical and timestamp stypes. Here is the implementation.

### CODE REVIEW

File: torch_frame/nn/encoding/positional_encoding.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            torch.arange(0, self.num_freqs, 2).float() *
+            (-np.log(10000.0) / self.num_freqs))
+        self.register_buffer("div_term", div_term)
+        position = torch.arange(self.num_freqs)
+        self.register_buffer("position", position)
+
+    def _forward(self, tensor: Tensor) -> Tensor:
+        assert torch.all(tensor >= 0)
+        positional_encoding = tensor.unsqueeze(-1) * self.position.reshape(
+            (1, ) * tensor.ndim + (-1, ))
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/nn/encoding/positional_encoding.py:31]: Also, you probably need to repeat `mult_term` for sine and cosine, respectively.

**Review by weihua916 — COMMENTED** [review]
Thank you. I think the implementation is different from the original paper. The pos in the original paper should be `input` in our case.
Screenshot 2023-11-29 at 7 01 15 PM

File: torch_frame/nn/encoding/positional_encoding.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            :obj:`Tensor`.
+
+    """
+    def __init__(self, num_freqs: int = 8):
+        super().__init__()
+        self.num_freqs = num_freqs
+        div_term = torch.exp(
+            torch.arange(0, self.num_freqs, 2).float() *
+            (-np.log(10000.0) / self.num_freqs))
+        self.register_buffer("div_term", div_term)
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/nn/encoding/positional_encoding.py:24]: I think it should be 
```python
torch.exp(-np.log(10000.0) * torch.arange(0, self.num_freqs, 2).float() / self.d_model)
```
`d_model` is an additional args in init

File: torch_frame/nn/encoding/positional_encoding.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    r"""Positional encoding introduced in `"Attention Is All You Need"
+    <https://arxiv.org/abs/1706.03762>`_ paper.
+
+    Args:
+        num_freqs (int): Number of frequencies to be applied to the
+            input :obj:`Tensor`, which is also the dimension of the output
+            :obj:`Tensor`.
+
+    """
+    def __init__(self, num_freqs: int = 8):
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/nn/encoding/positional_encoding.py:18]: and I'd just raise error if it is not divisible by 2

**Comment by weihua916** [comment]
@yiweny Made some fixes to align with the paper's formula.

**Comment by yiweny** [comment]
> @yiweny Made some fixes to align with the paper's formula. Also, I don't think we need to raise NaN error here. We already have a safeguard [here](https://github.com/pyg-team/pytorch-frame/blob/master/torch_frame/nn/encoder/stype_encoder.py#L123).

Yeah but I think you have no guarantee that it will only be called inside an encoder now that you put it into a module.

**Review by weihua916 — APPROVED** [review]

File: torch_frame/nn/encoding/positional_encoding.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    Args:
+        out_size (int): The output dimension size.
+    """
+    def __init__(self, out_size: int):
+        super().__init__()
+        if out_size % 2 != 0:
+            raise ValueError(
+                f"out_size should be divisible by 2 (got {out_size}).")
+        self.out_size = out_size
+        mult_term = torch.exp(-np.log(10000.0) *
```
Comment (zechengz) [inline torch_frame/nn/encoding/positional_encoding.py:22]: Why we use np.log here. Change to torch?

Comment (weihua916) [reply torch_frame/nn/encoding/positional_encoding.py:22]: oops

**Pull request merged**